src/fhirtypes/common.ts

```diff
@@ -43,12 +43,12 @@ export function setPropertyOnInstance(
         if (current[key] == null) current[key] = [];
         // If the index doesn't exist in the array, add it and lesser indices
         // Empty elements should be null, not undefined, according to https://www.hl7.org/fhir/json.html#primitive
-        let j = current[key].length;
-        for (; j < index; j++) {
-          current[key].push(null);
-        }
-        if (j === index) {
-          current[key].push({});
+        for (let j = 0; j <= index; j++) {
+          if (j < current[key].length && j === index && !current[key][index]) {
```

Comment:
I'm trying to work out if `!current[key][index]` is a safe way to do this check or if you need to use `current[key][index] != null`...  The current form would be a problem if it's possible for `current[key][index]` to have the value `false` or `0`.  Is that possible?

Since `j < current[key].length`, then I think we know we're not at the _end_ of the path.  So we can assume more comes after -- which means `false` or `0` wouldn't be valid.  Even in the case of children on primitives, if the real primitive value was set to `false`, then `current[key][index]` still wouldn't be pointing to it, because it would be pointing to the key with the `_` prefix, right?

I think that's right -- just want to confirm with you.

Reply:
It looks like we don't have any tests that set a real primitive value _and_ primitive children at the same time.  So, I think it would be good to actually add a test for this scenario anyway.  Basically it should be testing for this:
```
active = false
active.id = "foo"
```

Reply:
I think we actually could have a case where `current[key][index]` is `false`. The `j < current[key].length` is testing if we are at the end of the current array, not the end of the path. So if we had an array of boolean primitives that looked like `arr = [false, true, true]`, and we tried to set `arr[0] = true`, we would  mistakenly think `arr[0]` is not set, and set it to `{}`. Now, later on in the code, we would actually overwrite that `{}` and change it to `true`. But even though it works it isn't really "right", so I am going to change it to just `current[key][index] == null`.

Reply:
And I'll add that test.

Reply:
Ah, OK.  I misread the code.  Good call.  Thanks!